### PR TITLE
Kafka Source

### PR DESCRIPTION
**Pull request by tdas**

### CODE REVIEW

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  override def afterAll(): Unit = {
+    if (testUtils != null) {
+      testUtils.teardown()
+      testUtils = null
+      super.afterAll()
+    }
+  }
+
+  test("basic receiving") {
```
Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: maybe try this test case:

``` scala
  (1 to 50).foreach  { i =>
    test(s"basic receiving #$i") {
      val topic = s"topic$i"
      testUtils.createTopic(topic)

      val kafkaSource = KafkaSource(Set(topic), kafkaParams)
      val mapped =
        kafkaSource
          .toDS()
          .map(kv => new String(kv._2).toInt + 1)

      testStream(mapped)(
        AddKafkaData(kafkaSource, topic, 1, 2, 3),
        CheckAnswer(2, 3, 4),
        StopStream,
        DropBatches(1),
        StartStream,
        CheckAnswer(2, 3, 4),
        StopStream,
        AddKafkaData(kafkaSource, topic, 4, 5, 6),
        StartStream,
        CheckAnswer(2, 3, 4, 5, 6, 7),
        AddKafkaData(kafkaSource, topic, 7),
        CheckAnswer(2, 3, 4, 5, 6, 7, 8))
    }
  }
```

Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: Oh yeah, I am going to add more tests! That's why WIP. Going to add multi-topic tests, etc.

Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: Of course, I just tried that specific test and it uncovered several concurrency issues.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala [hunk 2]
```diff
@@ -35,19 +37,30 @@ case class CompositeOffset(offsets: Seq[Option[Offset]]) extends Offset {
...
-      if (signs.size != 1) {
-        throw new IllegalArgumentException(
-          s"Invalid comparison between non-linear histories: $this <=> $other")
+      val nonZeroSigns = comparisons.map(sign).filter { _ != 0}.toSet
+      nonZeroSigns.size match {
+        case 0 => 0                       // if both empty or only 0s
+        case 1 => nonZeroSigns.head       // if there are only (0s and 1s) or (0s and -1s)
+        case _ =>                          // there are both 1s and -1s
+          throw new IllegalArgumentException(
+            s"Invalid comparison between non-linear histories: $this <=> $other")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala:46]: @marmbrus This is a bug here. Basically, `[0, 0] < [0, 1]` is correct, but the earlier code would give exception as  `signs = [0, -1]`

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/OffsetSuite.scala [hunk 3]
```diff
@@ -33,22 +33,68 @@ trait OffsetSuite extends SparkFunSuite {
...
       assert(one >= one)
       assert(one == one)
+      assert(two == two)
+      assert(one != two)
+      assert(two != one)
+    }
+  }
+
+  /** Creates test to check that non-equality comparisons throw exception. */
+  def compareInvalid(one: Offset, two: Offset): Unit = {
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/OffsetSuite.scala:43]: @marmbrus Added invalid offset unit tests as well to ensure greater coverage.

**Comment by tdas** [comment]
I want to add the randomized testing as well, but I dont want this to get blocked by that. Please start reviewing this.

**Comment by tdas** [comment]
@marmbrus Ping for review. 

Also Jenkins does not seem to be testing this. What do I do to force it?

**Comment by tdas** [comment]
Jenkins, test this please.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+import scala.util.Try
+
+import kafka.common.TopicAndPartition
+import kafka.serializer._
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.execution.streaming.{Batch, Offset, Source, StreamingRelation}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, Dataset, SQLContext}
```
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala:27]: Spark/Kafka ordering are inconsistent with the file above.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+
+/** An [[Offset]] for the [[KafkaSource]]. */
+private[kafka]
+case class KafkaSourceOffset(offsets: Map[TopicAndPartition, Long]) extends Offset {
```
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala:32]: This probably does not need to be a case class give you are implementing equality and the toString already.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        case _ => // there are both 1s and -1s
+          throw new IllegalArgumentException(
+            s"Invalid comparison between non-linear histories: $this <=> $other")
+      }
+
+    case _ =>
+      throw new IllegalArgumentException(s"Cannot compare $this <=> $other")
+  }
+
+  override def ==(other: Offset): Boolean = Try(compareTo(other) == 0).getOrElse(false)
```
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala]: I would override `equals` since I think that `==` delegates to that.

I also would not hide exceptions, but instead make the matching for `false` explicit.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+            } else if (a > b) {
+              1
+            } else {
+              0
+            }
+          case (None, _) =>
+            -1
+          case (_, None) =>
+            1
+          case _ =>
```
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala]: Is this case possible?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+private[kafka] object KafkaSourceOffset {
+
+  /** Returns [[KafkaSourceOffset]] from a Option[Offset]. */
+  def from(offsetOption: Option[Offset]): Option[KafkaSourceOffset] = {
+    offsetOption.map { offset =>
+      offset match {
+        case o: KafkaSourceOffset => o
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Invalid conversion from offset of ${offset.getClass} to $getClass")
```
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala]: This message is a little misleading since `getClass` will the return the name of the singleton.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala (hunk 5 above)
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala:32]: Is there a reason this isn't just `KafkaOffset`?  Will there be one for anything other than the `Source`?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    }
+  }
+
+  def toDS(): Dataset[(Array[Byte], Array[Byte])] = {
+    toDF.as[(Array[Byte], Array[Byte])]
+  }
+
+  def toDF(): DataFrame = {
+    new DataFrame(sqlContext, logicalPlan)
+  }
```
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala:169]: I would consider defining these in the tests since they are really only for testing.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+  /** Get latest offsets from Kafka. */
+  private def getLatestOffsets(): KafkaSourceOffset = {
+    val partitionLeaders = KafkaCluster.checkErrors(kc.findLeaders(topicAndPartitions))
+    val leadersAndOffsets = KafkaCluster.checkErrors(kc.getLatestLeaderOffsets(topicAndPartitions))
+    KafkaSourceOffset(leadersAndOffsets.map { x => (x._1, x._2.offset) })
+  }
+
+  /** Get the initial offsets from Kafka for the source to start from. */
+  private def getInitialOffsets(): KafkaSourceOffset = {
+    if (params.get("auto.offset.reset").map(_.toLowerCase) == Some("smallest")) {
```
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala:180]: typically we do all extraction / validation of the options in the provider and then the arguments to the source are typed (instead of passing the map into the class itself).  The nice thing about this is it centralizes the location for all configuration.

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        AddKafkaData(kafkaSource, topic, 4, 5, 6),    // Add data when stream is stopped
+        StartStream,
+        CheckAnswer(2, 3, 4, 5, 6, 7),                // Should get the added data
+        AddKafkaData(kafkaSource, topic, 7, 8),
+        CheckAnswer(2, 3, 4, 5, 6, 7, 8, 9)
+      )
+    }
+  }
+
+  compare(
```
Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: I would put these in their own suite since they are proper unit tests (while the other seem more like integration tests).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 12]
```diff
@@ -44,7 +44,7 @@ class StreamExecution(
   var batchRun = false
 
   /** Minimum amount of time in between the start of each batch. */
-  val minBatchTime = 10
+  val minBatchTime = 100
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Why this change?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Offset.scala [hunk 13]
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.sql.execution.streaming
 
+import scala.util.Try
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Offset.scala]: remove?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala [hunk 14]
```diff
@@ -35,19 +37,32 @@ case class CompositeOffset(offsets: Seq[Option[Offset]]) extends Offset {
...
     case i if i < 0 => -1
     case i if i == 0 => 0
     case i if i > 0 => 1
   }
 }
+
+object CompositeOffset {
+  /**
+   * Returns a [[CompositeOffset]] with a variable sequence of offsets.
+   * `nulls` in the sequence are converted to `None`s. */
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala]: wrap the trailing `*/`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala [hunk 15]
```diff
@@ -35,19 +37,32 @@ case class CompositeOffset(offsets: Seq[Option[Offset]]) extends Offset {
...
+        case _ =>                          // there are both 1s and -1s
+          throw new IllegalArgumentException(
+            s"Invalid comparison between non-linear histories: $this <=> $other")
       }
-      signs.head
     case _ =>
       throw new IllegalArgumentException(s"Cannot compare $this <=> $other")
   }
 
+  override def ==(other: Offset): Boolean = Try(compareTo(other) == 0).getOrElse(false)
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala]: I think this should be `equals` as well.  Note that means we need to override hashcode too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala [hunk 16]
```diff
@@ -35,19 +37,32 @@ case class CompositeOffset(offsets: Seq[Option[Offset]]) extends Offset {
         case (None, _) => -1
         case (_, None) => 1
        }
-      val signs = comparisons.map(sign).distinct
-      if (signs.size != 1) {
-        throw new IllegalArgumentException(
-          s"Invalid comparison between non-linear histories: $this <=> $other")
+      val nonZeroSigns = comparisons.map(sign).filter { _ != 0}.toSet
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala]: use `()` not `{}`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala [hunk 17]
```diff
@@ -35,19 +37,32 @@ case class CompositeOffset(offsets: Seq[Option[Offset]]) extends Offset {
...
+      nonZeroSigns.size match {
+        case 0 => 0                       // if both empty or only 0s
+        case 1 => nonZeroSigns.head       // if there are only (0s and 1s) or (0s and -1s)
+        case _ =>                          // there are both 1s and -1s
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala]: Nit: off by one space.

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+  import testImplicits._
+
+  private var testUtils: KafkaTestUtils = _
+  private var kafkaParams: Map[String, String] = _
+
+  override val streamingTimout = 10.seconds
+
+  case class AddKafkaData(
+      kafkaSource: KafkaSource,
+      topic: String, data: Int*)(implicit multiPartitionCheck: Boolean = true) extends AddData {
```
Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: Its a little odd to make a `Boolean` an `implicit` parameter.  Also is this used?

Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: Nit: wrap `data` too.

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+          val maxOffsetInPartition = msgAndMetadata.map(_._2).maxBy(_.offset).offset()
+          (TopicAndPartition(topic, partition), maxOffsetInPartition + 1)
+        }
+
+      def metadataToStr(m: (String, RecordMetadata)): String = {
+        s"Sent ${m._1} to partition ${m._2.partition()}, offset ${m._2.offset()}"
+      }
+
+      assert(latestOffsetMap.size > 1,
+        s"Added data does not test multiple partitions: " + sentMetadata.map(metadataToStr))
```
Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: What does this mean?  Should we test this?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+
+    val offsetRanges = beginOffset to latestOffset
+    val kafkaParams = params
+    val encodingFunc = encoder.toRow _
+    val sparkContext = sqlContext.sparkContext
+
+    if (offsetRanges.nonEmpty) {
+      val rdd = KafkaUtils.createRDD[Array[Byte], Array[Byte], DefaultDecoder, DefaultDecoder](
+        sparkContext, kafkaParams, offsetRanges.toArray)
+      logInfo("Creating DF with offset ranges: " + offsetRanges)
```
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala]: Nit: Use string interpolation

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala (hunk 5 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala:32]: The package has other classes like OffsetRange etc that are used for Spark Streaming stuff. I thought its better differentiate this. Also, since Kafka's record id is also called offset, KafkaOffset sounds like the offset for a single topic+partition. So I think its less ambiguous to distinguish it by naming it `KafkaSourceOffset`.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala (hunk 7 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala]: Not possible. I can remove it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 12 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Removed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Offset.scala (hunk 13 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Offset.scala]: Removed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala (hunk 14 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala]: Done.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala (hunk 6 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala]: `Offset` class overrides `==`, hence i overrode this, just to make sure that this does not throw exception. 
Since we want to reuse the implementation of compareTo in both == and equals, it may be better to not define `==` in `Offset` but override `equals`. Implementation of `Offset` will have to ensure that equals never throws exception.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala (hunk 8 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala]: Changed to `KafkaSourceOffset`

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala (hunk 9 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala:169]: But MemoryStream has them defined in the class.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala (hunk 10 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala:180]: I can fix that while building the provider.

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala (hunk 18 above)
Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: Will be used for building stream tester, but in a different PR. If it bothers too much, then I can remove it from this PR.

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala (hunk 19 above)
Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: This is to make sure that data is added such that it gets added to multiple Kafka partitions. This will throw an exception if some one does not give data such that it gets spread across partitions.

Just to make the test future proof, so that someone accidentally does not disable multi-partition testing. Probably better to make this a `require` instead of an `assert`.

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala (hunk 11 above)
Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: Added to KafkSourceOffsetSuite

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala (hunk 18 above)
Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: I guess I prefer to remove anything that is not tested.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Offset.scala [hunk 21]
```diff
@@ -33,5 +35,17 @@ trait Offset extends Serializable {
   def <(other: Offset): Boolean = compareTo(other) < 0
   def <=(other: Offset): Boolean = compareTo(other) <= 0
   def >=(other: Offset): Boolean = compareTo(other) >= 0
-  def ==(other: Offset): Boolean = compareTo(other) == 0
+
+  override def equals(other: Any): Boolean = {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Offset.scala]: I'm not sure that its valid to override `equals` without also overriding `hashCode`

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+    for (i <- 0 until 5) {
+      logInfo(s"Round $i")
+
+      // Create Kafka source that reads from latest offset
+      val kafkaSource = KafkaSource(Set(topic), kafkaParams)
+
+      val mapped =
+        kafkaSource
+          .toDS()
+          .map[Int]((kv: (Array[Byte], Array[Byte])) => new String(kv._2).toInt + 1)
```
Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: I don't think that you need to specify `[Int]` here.

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+import kafka.common.TopicAndPartition
+import org.apache.kafka.clients.producer.RecordMetadata
+import org.scalatest.time.SpanSugar._
+
+import org.apache.spark.sql.StreamTest
+import org.apache.spark.sql.execution.streaming.{Offset, Source}
+import org.apache.spark.sql.test.SharedSQLContext
+
+
+class KafkaSourceSuite extends StreamTest with SharedSQLContext {
```
Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala:30]: I can't seem to get this to pass:

``` scala
  test("stress test") {
    val topic = "stress"
    testUtils.createTopic(topic, partitions = 2)

    // Create Kafka source that reads from latest offset
    val kafkaSource = KafkaSource(Set(topic), kafkaParams)

    val mapped =
      kafkaSource
          .toDS()
          .map(r => new String(r._2).toInt + 1)

    createStressTest(mapped, d => AddKafkaData(kafkaSource, "stress", d: _*))
  }
```

Am I doing something wrong?

Comment (marmbrus) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala:30]: In particular, I think this fails when you have a batch where there is a "quiet" stream. (which is currently hard to write a test for given the `require` below).

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala (hunk 22 above)
Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala]: Removed

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala (hunk 23 above)
Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala:30]: Thats why i had added the extra implicit param to disable the require() for the stress test i was going to add. I was going to add a more complicated stress test. I can add this for now, but I am going to change this in a later PR.

Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala:30]: The problem was `AddKafkaData` did not return correct offsets if there no data was actually added, as it used the offsets returned on every successful push. Changed that. 

And adding the more complicated stress test as well with multiple topics.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala (hunk 9 above)
Comment (marmbrus) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala:169]: Okay, we should consider moving these once we have a real provider for this Source.
